test/src/test/java/aima/test/unit/search/basic/BasicPriorityFrontierQueueTest.java

```diff
@@ -0,0 +1,131 @@
+package aima.test.unit.search.basic;
+
+import aima.core.search.basic.support.BasicFrontierQueue;
+import aima.core.search.basic.support.BasicNodeFactory;
+import aima.core.search.basic.support.BasicPriorityFrontierQueue;
+import aima.core.search.api.Node;
+import aima.core.search.api.NodeFactory;
+
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.PriorityQueue;
+
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+
+public class BasicPriorityFrontierQueueTest<A, S> {
+	
+	private static NodeFactory<Integer, String> nodeFactory;
+	private static int NUMBER_OF_NODES;
+	private static int NUMBER_OF_STATES;
+	private static Node[] testNodes;
+	private static Node[] fastPQResult;
+	private static Node[] normalPQResult;
+	private static Comparator<Node<Integer, String>> comparator;
+	
+	@BeforeClass
+	public static void prepare() {
+		
+		
+		nodeFactory = new BasicNodeFactory<Integer, String>();
+		
+		//Custom comparator to avoid duplicates in a TreeSet
+		//Checks on pathCost first, followed by state (only for a clearer output), and finally followed by object equality
+		
+		comparator = (Node<Integer, String> node1, Node<Integer, String> node2)-> 
+					 ((node1.pathCost() == node2.pathCost()) ? ((node1.state().equals(node2.state())) ? (node1.equals(node2) ? 0 : 1) : (node1.state().compareTo(node2.state()))) 
+					 : (int)Math.signum((node1.pathCost() - node2.pathCost())));
+		
+	    // Small Test Cases for fastPQ

```

Comment:
However, the comparator and the fastQ do not work well together, odd behavior, try:

for (int i = 0; i < 10; i++) {			
 	System.out.println(i+" = "+ fastPQ.size());
	fastPQ.add(node);
}

and compare with a normal priority queue and you'll see it does not work as expected.

This needs to be rethought, as the comparator is too complex to expect a user of the API to setup correctly. An approach may be inside of BasicPriorityFrontierQueue would be to create a NodeContainer object that holds onto an insertion id (that auto increments and doesn't repeat) which is taken into account in the internal 'queue's comparator but not the 'stateNodeMap' comparator.